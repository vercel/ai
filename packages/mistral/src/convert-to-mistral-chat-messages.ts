import {
  LanguageModelV1Prompt,
  UnsupportedFunctionalityError,
} from '@ai-sdk/provider';
import { convertUint8ArrayToBase64 } from '@ai-sdk/provider-utils';
import { MistralPrompt } from './mistral-chat-prompt';

export function convertToMistralChatMessages(
  prompt: LanguageModelV1Prompt,
): MistralPrompt {
  const messages: MistralPrompt = [];

  for (let i = 0; i < prompt.length; i++) {
    const { role, content } = prompt[i];
    const isLastMessage = i === prompt.length - 1;

    switch (role) {
      case 'system': {
        messages.push({ role: 'system', content });
        break;
      }

      case 'user': {
        messages.push({
          role: 'user',
          content: content.map(part => {
            switch (part.type) {
              case 'text': {
                return { type: 'text', text: part.text };
              }
              case 'image': {
                return {
                  type: 'image_url',
                  image_url:
                    part.image instanceof URL
                      ? part.image.toString()
                      : `data:${
                          part.mimeType ?? 'image/jpeg'
                        };base64,${convertUint8ArrayToBase64(part.image)}`,
                };
              }
              case 'file': {
                if (!(part.data instanceof URL)) {
                  throw new UnsupportedFunctionalityError({
                    functionality: 'File content parts in user messages',
                  });
                }

                switch (part.mimeType) {
                  case 'application/pdf': {
                    return {
                      type: 'document_url',
                      document_url: part.data.toString(),
                    };
                  }
                  default: {
                    throw new UnsupportedFunctionalityError({
                      functionality:
                        'Only PDF files are supported in user messages',
                    });
                  }
                }
              }
            }
          }),
        });
        break;
      }

      case 'assistant': {
        let text = '';
        const toolCalls: Array<{
          id: string;
          type: 'function';
          function: { name: string; arguments: string };
        }> = [];

        for (const part of content) {
          switch (part.type) {
            case 'text': {
              text += part.text;
              break;
            }
            case 'tool-call': {
              toolCalls.push({
                id: part.toolCallId,
                type: 'function',
                function: {
                  name: part.toolName,
                  arguments: JSON.stringify(part.args),
                },
              });
              break;
            }
          }
        }

        messages.push({
          role: 'assistant',
          content: text,
          prefix: isLastMessage ? true : undefined,
          tool_calls: toolCalls.length > 0 ? toolCalls : undefined,
        });

        break;
      }
      case 'tool': {
        for (const toolResponse of content) {
          messages.push({
            role: 'tool',
            name: toolResponse.toolName,
            content: JSON.stringify(toolResponse.result),
            tool_call_id: toolResponse.toolCallId,
          });
        }
        break;
      }
      default: {
        const _exhaustiveCheck: never = role;
        throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
      }
    }
  }

  return messages;
}
