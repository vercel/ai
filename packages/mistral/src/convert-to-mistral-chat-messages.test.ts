import { convertToMistralChatMessages } from './convert-to-mistral-chat-messages';

describe('user messages', () => {
  it('should convert messages with image parts', async () => {
    const result = convertToMistralChatMessages([
      {
        role: 'user',
        content: [
          { type: 'text', text: 'Hello' },
          {
            type: 'file',
            data: 'AAECAw==',
            mediaType: 'image/png',
          },
        ],
      },
    ]);

    expect(result).toMatchSnapshot();
  });

  it('should convert messages with PDF file parts using URL', () => {
    const result = convertToMistralChatMessages([
      {
        role: 'user',
        content: [
          { type: 'text', text: 'Please analyze this document' },
          {
            type: 'file',
            data: new URL('https://example.com/document.pdf'),
            mediaType: 'application/pdf',
          },
        ],
      },
    ]);

    expect(result).toMatchSnapshot();
  });
});

describe('tool calls', () => {
  it('should stringify arguments to tool calls', () => {
    const result = convertToMistralChatMessages([
      {
        role: 'assistant',
        content: [
          {
            type: 'tool-call',
            input: { key: 'arg-value' },
            toolCallId: 'tool-call-id-1',
            toolName: 'tool-1',
          },
        ],
      },
      {
        role: 'tool',
        content: [
          {
            type: 'tool-result',
            toolCallId: 'tool-call-id-1',
            toolName: 'tool-1',
            output: { type: 'json', value: { key: 'result-value' } },
          },
        ],
      },
    ]);

    expect(result).toMatchInlineSnapshot(`
      [
        {
          "content": "",
          "prefix": undefined,
          "role": "assistant",
          "tool_calls": [
            {
              "function": {
                "arguments": "{"key":"arg-value"}",
                "name": "tool-1",
              },
              "id": "tool-call-id-1",
              "type": "function",
            },
          ],
        },
        {
          "content": "{"key":"result-value"}",
          "name": "tool-1",
          "role": "tool",
          "tool_call_id": "tool-call-id-1",
        },
      ]
    `);
  });

  it('should handle text output format', () => {
    const result = convertToMistralChatMessages([
      {
        role: 'assistant',
        content: [
          {
            type: 'tool-call',
            input: { query: 'test' },
            toolCallId: 'tool-call-id-2',
            toolName: 'text-tool',
          },
        ],
      },
      {
        role: 'tool',
        content: [
          {
            type: 'tool-result',
            toolCallId: 'tool-call-id-2',
            toolName: 'text-tool',
            output: { type: 'text', value: 'This is a text response' },
          },
        ],
      },
    ]);

    expect(result).toMatchInlineSnapshot(`
      [
        {
          "content": "",
          "prefix": undefined,
          "role": "assistant",
          "tool_calls": [
            {
              "function": {
                "arguments": "{"query":"test"}",
                "name": "text-tool",
              },
              "id": "tool-call-id-2",
              "type": "function",
            },
          ],
        },
        {
          "content": "This is a text response",
          "name": "text-tool",
          "role": "tool",
          "tool_call_id": "tool-call-id-2",
        },
      ]
    `);
  });

  it('should handle content output format', () => {
    const result = convertToMistralChatMessages([
      {
        role: 'assistant',
        content: [
          {
            type: 'tool-call',
            input: { query: 'generate image' },
            toolCallId: 'tool-call-id-3',
            toolName: 'image-tool',
          },
        ],
      },
      {
        role: 'tool',
        content: [
          {
            type: 'tool-result',
            toolCallId: 'tool-call-id-3',
            toolName: 'image-tool',
            output: {
              type: 'content',
              value: [
                { type: 'text', text: 'Here is the result:' },
                { type: 'media', data: 'base64data', mediaType: 'image/png' },
              ],
            },
          },
        ],
      },
    ]);

    expect(result).toMatchInlineSnapshot(`
      [
        {
          "content": "",
          "prefix": undefined,
          "role": "assistant",
          "tool_calls": [
            {
              "function": {
                "arguments": "{"query":"generate image"}",
                "name": "image-tool",
              },
              "id": "tool-call-id-3",
              "type": "function",
            },
          ],
        },
        {
          "content": "[{"type":"text","text":"Here is the result:"},{"type":"media","data":"base64data","mediaType":"image/png"}]",
          "name": "image-tool",
          "role": "tool",
          "tool_call_id": "tool-call-id-3",
        },
      ]
    `);
  });

  it('should handle error output format', () => {
    const result = convertToMistralChatMessages([
      {
        role: 'assistant',
        content: [
          {
            type: 'tool-call',
            input: { query: 'test' },
            toolCallId: 'tool-call-id-4',
            toolName: 'error-tool',
          },
        ],
      },
      {
        role: 'tool',
        content: [
          {
            type: 'tool-result',
            toolCallId: 'tool-call-id-4',
            toolName: 'error-tool',
            output: { type: 'error-text', value: 'Invalid input provided' },
          },
        ],
      },
    ]);

    expect(result).toMatchInlineSnapshot(`
      [
        {
          "content": "",
          "prefix": undefined,
          "role": "assistant",
          "tool_calls": [
            {
              "function": {
                "arguments": "{"query":"test"}",
                "name": "error-tool",
              },
              "id": "tool-call-id-4",
              "type": "function",
            },
          ],
        },
        {
          "content": "Invalid input provided",
          "name": "error-tool",
          "role": "tool",
          "tool_call_id": "tool-call-id-4",
        },
      ]
    `);
  });
});

describe('assistant messages', () => {
  it('should add prefix true to trailing assistant messages', () => {
    const result = convertToMistralChatMessages([
      {
        role: 'user',
        content: [{ type: 'text', text: 'Hello' }],
      },
      {
        role: 'assistant',
        content: [{ type: 'text', text: 'Hello!' }],
      },
    ]);

    expect(result).toMatchSnapshot();
  });
});
