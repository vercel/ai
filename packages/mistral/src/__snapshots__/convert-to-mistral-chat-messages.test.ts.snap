// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assistant messages > should add prefix true to trailing assistant messages 1`] = `
[
  {
    "content": [
      {
        "text": "Hello",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "Hello!",
    "prefix": true,
    "role": "assistant",
    "tool_calls": undefined,
  },
]
`;

exports[`user messages > should convert messages with PDF file parts using URL 1`] = `
[
  {
    "content": [
      {
        "text": "Please analyze this document",
        "type": "text",
      },
      {
        "document_url": "https://example.com/document.pdf",
        "type": "document_url",
      },
    ],
    "role": "user",
  },
]
`;

exports[`user messages > should convert messages with image parts 1`] = `
[
  {
    "content": [
      {
        "text": "Hello",
        "type": "text",
      },
      {
        "image_url": "data:image/png;base64,AAECAw==",
        "type": "image_url",
      },
    ],
    "role": "user",
  },
]
`;
